File: monitor/src/main/java/com/caibowen/prma/monitor/eval/Evaluator.java
package com.caibowen.prma.monitor.eval;

import com.caibowen.gplume.annotation.Const;
import com.caibowen.prma.api.model.EventVO;

import javax.annotation.Nonnull;

/**
 * @author BowenCai
 * @since 5-11-2014.
 */
public interface Evaluator {

    public static final String ACCEPT = Evaluator.class.getName() + ".ACCEPT";
    public static final String REJECT = Evaluator.class.getName() + ".REJECT";
    /**
     * TODO
     * 1. store cache
     * 2. java code block eval integrate
     * 3. monitor eval chain building
     * 4. notify stmp, http post
     * 5. gplume bean namespace support -> OK
     * 6. context build
     * 7. analyzing API
     *
     *
     * optional :
     * 1. bean assembler <required></required> -> pending
     * 2. bean proxy support -> OK
     * 3. socket logger
     * 4. distributed logging system demo
     */
    String eval(@Nonnull @Const EventVO vo);
}

File: monitor/src/main/scala/com/caibowen/prma/monitor/eval/Evaluator.scala
package com.caibowen.prma.monitor.eval

import com.caibowen.prma.api.model.EventVO


/**
 * @author BowenCai
 * @since  30/11/2014.
 */

class Response(val event: EventVO)
case object Reject extends Response(null)
class NotifyOne(val name: String, event: EventVO) extends Response(event)

// response that carries extra data for the notifier
trait WithData[T] {val data: T}

class NotifyAllWith[T](val data: T, event: EventVO) extends Response(event) with WithData[T]

class NotifyWith[T](name: String, val data: T, event: EventVO) extends NotifyOne(name, event) with WithData[T]

trait Evaluator {
  def eval(vo : EventVO): Response = Reject
}
